### PR TITLE
More general parser

### LINKED ISSUE #40: Binary Coding
This causes an error
https://tuwien2020.github.io/tgi-pages/#/binary-coding?input-offset=

### LINKED ISSUE #44: Binary Calculator
Route: binary-calculator

Division cuts off after 8 decimal places

### PR DESCRIPTION
**Pull request by stefnotch**
I'm refactoring the parsing code so that we only have one, well documented parser. That parser outputs a rather generic format. For example `a xor truE` would be parsed as `["Xor", "a", "truE"]`

Then, we have 'transformers' that take that generic format and can output a slightly more specialized one. e.g. `["Xor", "a", true]`

And finally, we have 'custom printers' that know how to print those specialized things

Additionally, the calculator now lets you input how many digits of the result it should show. And one bug on the binary coding page has been fixed

fix #40 
fix #44

### CODE REVIEW

**Comment by stefnotch** [comment]
Sweet, so far it looks like this is going to *simplify* the code
image

**Comment by stefnotch** [comment]
Aw yiss, better error reporting. Not that we needed it, but it's a cute side-effect of the new architecture.
image

**Comment by stefnotch** [comment]
To the reviewer: I apologize for having written this much code. Please do test the interface and see if you can get it to crash or anything

File: src/math/binary-expression.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    for (let i = 0; i < chunks.length; i++) {
+      const position = i * chunkSize + startOffset;
+      chunks[i] = value.substring(position, position + chunkSize);
+    }
+
+    return chunks;
+  }
+
+  const mathJsonOperatorMap = new Map<MathJsonMathOperator, (a: BinaryNumber, b: BinaryNumber) => BinaryNumber>([
```
Comment (Fancy11111) [inline src/math/binary-expression.ts:143]: can you explain to me the purpose of this map, because I don't really understand what it is supposed to do.

Comment (Fancy11111) [reply src/math/binary-expression.ts:143]: I guess the values are irrelevant, but why would you use a complex type for the value then

File: src/math/boolean-expression.ts [hunk 2]
```diff
@@ -47,19 +19,62 @@ export function useBooleanExpressions() {
...
-    ["not", (a) => !a],
-    ["implies", (a, b) => !a || b],
-    ["and", (a, b) => a && b],
-    ["or", (a, b) => a || b],
-    ["xor", (a, b) => (a ? !b : b)],
-    ["nand", (a, b) => !(a && b)],
-    ["nor", (a, b) => !(a || b)],
-    ["equals", (a, b) => a == b],
-    ["subset", (a, b) => !b || a],
+    ["Not", (a) => !a],
```
Comment (Fancy11111) [inline src/math/boolean-expression.ts:66]: General question, was there a reason for the changing of all operator names from all lowercase to first letter uppercase?

Comment (stefnotch) [reply src/math/boolean-expression.ts:66]: Good question, I suppose this is worth documenting. The reason for doing so was that I mimicked this MathJson https://cortexjs.io/math-json/

File: src/math/binary-expression.ts (hunk 1 above)
Comment (stefnotch) [reply src/math/binary-expression.ts:143]: The idea was that one could give it some MathJson and this map would let you calculate the result. Though, I suppose I'll scrap those complicated types for now

File: src/components/MathOutput.vue [hunk 3]
```diff
@@ -1,50 +1,77 @@
...
       type: [Object, String, Array, Boolean],
-      required: true,
+      required: false,
+    },
+    latex: {
+      type: String,
+      required: false,
+    },
+    formatting: {
+      type: Object,
```
[the remark this answers is not in the document]
Comment (stefnotch) [reply src/components/MathOutput.vue:23]: Oh, I forgot that I could do that. 😅

File: src/pages/BinaryCalculator.vue [hunk 4]
```diff
@@ -1,48 +1,37 @@
...
-  }
-
-  return {
-    parseBinaryOperation,
-  };
-}
+import { BinaryNumber } from "../math/binary-number";
 
 export default defineComponent({
   components: { MathInput, MathOutput, MathSingleOp },
```
[the remark this answers is not in the document]
Comment (stefnotch) [reply src/pages/BinaryCalculator.vue:37]: Thanks for pointing that out

**Comment by stefnotch** [comment]
@Fancy11111 Thank you very much for actually taking your time to review this! Super appreciated

**Review by Fancy11111 — APPROVED** [review]

**Pull request merged**